## Introduction 👋👋👋
___

We're about to dive into the theory of what convents are why they have been so successful at computer vision tasks.But first, let's take a pratical look at simple convnet example.It uses a convnet to classify MNIST digits,Even though convnet will be basic, it accuracy will blow out of the water that of the densely connected model.

Convents architecure are basically stack of `Conv2D` and `MaxPooling2D` layers.

### Instantiating a small convnet
___
#### Dataset preparation 🗃️🗃️🗃️

In [1]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt

In [12]:
(x_train,y_train),(x_test,y_test) = keras.datasets.mnist.load_data()

11493376/11490434 [==============================] - 1s 0us/step


In [13]:
x_train.shape, y_train.shape

((60000, 28, 28), (60000,))

In [14]:
x_test.shape, y_test.shape

((10000, 28, 28), (10000,))

In [16]:
x_train.min()

0

In [17]:
x_train.max()

255

#### Normalizing train data  ✂️✂️✂️

In [18]:
x_train = x_train /255
x_test = x_test /255

In [19]:
x_train.min()

0.0

In [20]:
x_train.max()

1.0

### label categorization
___

In [21]:
y_train = keras.utils.to_categorical(y_train)
y_test = keras.utils.to_categorical(y_test)

In [24]:
y_train.shape,y_test.shape

((60000, 10), (10000, 10))

### Building convnet model 🏗️🏗️🏗️
___

In [69]:
from tensorflow.keras.layers import Dense,Conv2D,MaxPooling2D
from tensorflow.keras import models
from tensorflow import keras

In [70]:
model = models.Sequential()
model.add(Conv2D(filters=32,kernel_size=(3,3),input_shape=(28,28,1),activation="relu",))
model.add(MaxPooling2D((2,2)))
model.add(Conv2D(filters=64,kernel_size=(3,3),activation="relu"))
model.add(MaxPooling2D((2,2)))
model.add(Conv2D(filters=64,kernel_size=(3,3),activation="relu"))

A convnet takes as input tensor of shape `(image_height,image_width,image_channels)`(not including batch dimension).In this case,we'll configure the convnet to process the input of size `(28,28,1)`, which is the format of MNIST images

In [71]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_6 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_7 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_8 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


 As you can see that output of every `Conv2D` and `MaxPooling2D` layer is a 3D tensor od shape `(height,width,channels)`.The width and height dimensions tend to shrink as you go deeper in the network.The number of channels is controlled by the first argument passed to the `Conv2D` layers(32 or 64). 
 
 The next step is to feed the last ouput tensor (of shape`(3,3,64)`) into a densely connected classifier network like those which we have implemented earlier: A stack of `Dense layers`.These classifiers process vectors, which are 1D, where the current ouput is a 3D tensor.First we have to flatten the 3D outputs to 1D, and then add a few `Dense` layers on top.

### Adding a classifier on top of the convnet
___

In [72]:
model.add(keras.layers.Flatten())
model.add(keras.layers.Dense(64,activation="relu"))
model.add(keras.layers.Dense(10,activation="softmax"))

In [73]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_6 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_7 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_8 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_7 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 64)               

As you can see, the (3,3,64) outputs are flattend into vectors of shape(576,) before going through `two Dense Layers`.